doc, src: sort cli cmd options and env vars

willhayslett opened the pull request:
Alphabetizing the command line options and environment
variables in doc/api/cli.md for consistency and readability.

Refs: https://github.com/nodejs/node/issues/19814

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

vsemozhetbyt commented:
Tricky diff)

vsemozhetbyt commented:
Let's do the check in some steps.

1. Sort all lines in both files, compare the result and make sure nothig unwanted is added/deleted.

Now:

* \+ 14 blank lines: more consistent space between sections, it is good.
* Metadata version statments changes seems accidental?

```diff
### `--force-fips`
<!-- YAML
- added: v6.0.0
 -->

### `--force-fips`
<!-- YAML
+added:
+  - v5.0.0
+  - v4.2.0
+changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/19600
+    description: The `--require` option is now supported when checking a file.
 -->
```

```diff
### `-c`, `--check`
<!-- YAML
added:
  - v5.0.0
  - v4.2.0
changes:
-  - version: REPLACEME
-    pr-url: https://github.com/nodejs/node/pull/19600
-    description: The `--require` option is now supported when checking a file.
 -->

### `-c`, `--check`
<!-- YAML
added:
  - v5.0.0
  - v4.2.0
 -->
```

vsemozhetbyt commented:
While fixing the metadata change, you can also add additional blank line between `` ### `NODE_NO_WARNINGS=1` `` and `` ### `NODE_OPTIONS=options...` `` sections for consistensy.

willhayslett commented:
Yes, very tricky diff; thank you for the thorough review!

I got a little tripped up on the ```force-fips``` metadata changes. Encountered a conflict with the following sha while rebasing:

https://github.com/nodejs/node/commit/0876a0314d15129e8b29cb6e9753b886086cbfe2#diff-85b6573bc4afd5e06ed3b9d9b553635d

Looks like the `force-fips` change is correct, but the removal from `--check` isn't.

I'll go ahead get this updated along with the blank lines, but let me know if that doesn't seem correct.

vsemozhetbyt commented:
It seems the removal from `--check` isn't correct, but I doubt if `--force-fips` change is correct: the changed and added metadata seems just copied from the `--check` change. Maybe `--force-fips` change should be reverted and `--check` metadata updated (during rebase)?

willhayslett commented:
@vsemozhetbyt, you're right. Making these changes now.

vsemozhetbyt commented:
[x] Sort all lines in temporal copies of both versions, compare the result and make sure nothing unwanted is added/deleted. — Check!

Now I am trying to find a way to check if nothing is changed inside the reordered sections by chance...

willhayslett commented:
@vsemozhetbyt it's definitely non-trivial to sift through the changes. 

I'm also working on making the same adjustments in ```node.cc``` so the options/vars are sorted when executing ```node --help```.  

If there happens to be a better way to do it so the diff isn't so difficult to read, please let me know.

vsemozhetbyt commented:
I am not aware of any simpler way(

But to check if the sections are not changed I've used this trick

1. Replace in both versions `/\n(?!#|\[.+\]:)/` with `' '` to make all the sections one-liners (except bottom references).
2. Trim trailing spaces.
3. Sort all lines.
4, The files should be equals.

And they are!

vsemozhetbyt commented:
Sorting is also OK.

vsemozhetbyt commented:
CI-lite: https://ci.nodejs.org/job/node-test-pull-request-lite/452/

vsemozhetbyt commented:
Changes in `node.cc` can be added to this PR as a separate commit.

willhayslett commented:
@vsemozhetbyt, as there were some logic changes with the latest commit, I'd like to take a stab at creating some c++ tests before this moves forward.

Also, noticed some cli options enumerated in ```node.cc``` that are not in ```doc/api/cli.md```. Specifically the following:

> --napi-modules 
    --v8-pool-size=num
    --experimental-modules
    --experimental-vm-modules

Should I add those to the doc as part of this PR or create a different one?

vsemozhetbyt commented:
I think as we are at this so deep and as we've already added some things from `cli.md` to `node.cc`, we can also do vice-versa. Just let's do it in a separate commit so that the first one remains reordering only, to not overcomplicate it.

willhayslett commented:
Got it, @vsemozhetbyt!

vsemozhetbyt commented:
So, I do not know C++, so the second commit needs to be reviewed by someone else. Maybe @Trott?
I just can find out this diffs resume with "sort-all-lines-and-compare" hack described above:
```diff
+         "  --                         indicate the end of node options\n"

+         "  -h, --help                 print node command line options\n"

+#if defined(NODE_HAVE_I18N_SUPPORT)
+         "NODE_PRESERVE_SYMLINKS       set to 1 to preserve symbolic links\n"
+         "                             when resolving and caching modules\n"
+#endif
```
(TBH, I do not know why `NODE_HAVE_I18N_SUPPORT` affects `NODE_PRESERVE_SYMLINKS`. Is this true?).
```diff
-         "  -                          script read from stdin (default; "
-         "interactive mode if a tty)\n"
+         "  -                          script read from stdin (default; \n"
+         "                             interactive mode if a tty)\n"

-    "--require", "-r",
+    "--require",
...
+    "-r",
```
And also these added and deleted lines in various places:
```dif
-#endif  /* NODE_FIPS_MODE */
-#endif /* HAVE_OPENSSL */
```
```diff
+#endif  // defined(NODE_HAVE_I18N_SUPPORT)
+#endif  // defined(NODE_HAVE_I18N_SUPPORT)
+#endif  // HAVE_INSPECTOR
+#endif  // HAVE_OPENSSL
+#endif  // HAVE_OPENSSL
+#endif  // HAVE_OPENSSL
+#endif  // HAVE_OPENSSL
+#endif  // NODE_FIPS_MODE
+#if defined(NODE_HAVE_I18N_SUPPORT)
+#if defined(NODE_HAVE_I18N_SUPPORT)
+#if HAVE_OPENSSL
+#if HAVE_OPENSSL
+#if HAVE_OPENSSL
```

vsemozhetbyt commented:
Intermediate full CI to check the current stage: https://ci.nodejs.org/job/node-test-pull-request/14131/

vsemozhetbyt commented:
One fail of `parallel/test-http2-stream-client` may be unrelated.

Trott commented:
> One fail of parallel/test-http2-stream-client may be unrelated.

Indeed, appears unrelated. Should re-run node-test-commit-linux just to set a good example if nothing else.  (Well, that and I think our current governance rules require it.) But I can't restart or add any node-test-commit-linux jobs right now. They just disappear in CI. @nodejs/build

rvagg commented:
I found a job (git-clean-rpi) that had a big queue cause old machines it was trying to run had been removed. I've fixed that up for now and new runs seem to be going through. I'm going to restart jenkins when I see it idle but for now it _seems_ to be working OK.

rvagg commented:
node-test-pull-request @ https://ci.nodejs.org/job/node-test-pull-request/14133/

I think you need to also edit doc/node.1 for completeness here, it has the same list of arguments in the same order I think. So you'd better get your roff on and figure that one out too (`man -l doc/node.1` to test it btw)

willhayslett commented:
haha yep @rvagg! Was going to ask if that needed to be done as well. Started looking at mdoc and roff... should be fun! Thanks for the test command!

willhayslett commented:
Do we have a way to perform tests against different versions of the compiled binary? I'd like to validate the latest commit against different build configurations. Specifically, the output of the ```PrintHelp()``` function changes based on the following build flags:

```
HAVE_OPENSSL: --without-ssl
NODE_FIPS_MODE: --openssl-fips
NODE_HAVE_I18N_SUPPORT: --with-icu-path
NODE_HAVE_SMALL_ICU: --with-intl; --without-intl
HAVE_INSPECTOR: --without-inspector
NODE_OPENSSL_CERT_STORE: --openssl-use-def-ca-store
NODE_WITHOUT_NODE_OPTIONS: --without-node-options
```
Is there any way to test against these with the current toolset?

Trott commented:
> Do we have a way to perform tests against different versions of the compiled binary?
[SNIP]
> Is there any way to test against these with the current toolset?

@willhayslett I _think_ the answer is "Unfortunately, not easily." I'm think you'd have to compile with the different options and test each resulting binary separately. Maybe move the binaries to a different directory so you can copy them back and forth as needed to avoid re-compiling? (And ccache will help with the compile time, if you don't have that installed already.)

However, in case there is a better way that I'm simply not aware of: @nodejs/build

rvagg commented:
we do a few in https://ci.nodejs.org/job/node-test-commit-linux-containered/

* debug
* fips
* shared openssl 1.0.2
* shred openssl 1.1.0
* without intl

others can be added as they are deemed a priority and our test suite can actually handle it.. `--without-ssl` is one I'd like to add but I bet we have a bunch of tests that don't do the proper checks

rvagg commented:
oh and node-test-commit-linux-containered is run as part of node-test-commit if that wasn't clear, so these are run each time

fips and 1.0.2 are going away in node 10 FYI so they aren't tested on current master

Trott commented:
@rvagg I *think* @willhayslett was looking for a solution they could run locally. They don't have CI access.

Trott commented:
(Although it's great to know that some of these compile options are tested in CI. I didn't know we did that! That's great!)

willhayslett commented:
Thanks, @rvagg and @Trott!

And yes, I was ultimately looking to create some unit tests for the changes. 

I did some manual testing and things look good, but I'd like to have something automated if possible. I'll look at what is being done on the CI side to see if I can leverage what's being done there at all.

Trott commented:
@willhayslett Any unit tests you write and put in `test/parallel` or `test/sequential` (or a few other directories, but those are the most common) will be run in CI every time a change is made to the code base. So the variants that @rvagg mentions will be tested automatically. So there's that! The downside is you are currently reliant on a Collaborator to start CI for you. So it may be kind of slow to iterate....

willhayslett commented:
Got it @Trott! 

Here's where we are:

1. `doc/api/cli.md`:
      * command line options sorted
      * additional formatting
      * added missing options
1. `src/node.cc`
      * command line options sorted
      * added missing options
1. `doc/api/node1`
      * command line options sorted

The c++ unit tests are something I'd like to do for posterity, but I'd prefer to study up a bit and get a better grasp of the language before I try to commit anything. 

Also, I'm still very curious about https://github.com/nodejs/node/pull/19878#issuecomment-379617870

Are `--preserve-symlinks`, `--experimental-modules` and `--experimental-vm-modules` only available  when compiled with `NODE_HAVE_I18N_SUPPORT`? Not sure what localization would have to do with those options.

willhayslett commented:
Hi @vsemozhetbyt or @Trott, can I get another CI run?

vsemozhetbyt commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14208/

willhayslett commented:
Thanks, @vsemozhetbyt!

vsemozhetbyt commented:
Do I get this right that we should land this after commit(s) with test(s)?

willhayslett commented:
If you're okay with landing without the tests, feel free. I'm hoping to get a chance to work on them tonight but not being familiar with c++, I'm not sure how long it'll take for me to get them done. There aren't a lot of examples of what I'm trying to do specifically. If I can finish before this lands then I'll add them here. Otherwise I can open another PR.

vsemozhetbyt commented:
Let's see what others think)

willhayslett commented:
@vsemozhetbyt looks like I can get the tests done in JS using some of the information available in the `common` and `process` objects 👍! Let's hold off on landing until I get them done, which will likely be sometime tomorrow evening.

willhayslett commented:
Hi @vsemozhetbyt, tests committed! I recompiled manually to test each of the options being asserted against and things looked good. Hopefully the CI tests reveal the same!

richardlau commented on test/parallel/test-cli-node-print-help.js:
```diff
@@ -0,0 +1,60 @@
+function testForSubstring(options) {
+  if (options.compileConstant) {
+    options.flags.forEach((flag) => {
+      assert.strictEqual(true, stdOut.indexOf(flag) !== -1);
```
Can you swap the arguments around? assert arguments are `actual, expected`.

willhayslett replied:
done

richardlau commented on test/parallel/test-cli-node-print-help.js:
```diff
@@ -0,0 +1,60 @@
+    });
+  } else {
+    options.flags.forEach((flag) => {
+      assert.strictEqual(-1, stdOut.indexOf(flag));
```
Can you swap the arguments around? assert arguments are `actual, expected`.

willhayslett replied:
done!

richardlau commented on test/parallel/test-cli-node-print-help.js:
```diff
@@ -0,0 +1,60 @@
+function startPrintHelpTest() {
+  exec('./node --help', common.mustCall((err, stdout, stderr) => {
+    if (err) {
+      assert.ifError(err);
```
This doesn't need to be wrapped in `if (err) {}`.

willhayslett replied:
done

richardlau commented on test/parallel/test-cli-node-print-help.js:
```diff
@@ -0,0 +1,60 @@
+
+
+function startPrintHelpTest() {
+  exec('./node --help', common.mustCall((err, stdout, stderr) => {
```
Maybe use `process.execPath`?

willhayslett replied:
done

richardlau commented:
Test failing on Windows, e.g., https://ci.nodejs.org/job/node-test-binary-windows/16492/COMPILED_BY=vs2017,RUNNER=win10,RUN_SUBSET=1/console:

```
not ok 64 parallel/test-cli-node-print-help
  ---
  duration_ms: 0.140
  severity: fail
  exitcode: 1
  stack: |-
    assert.js:589
        throw newErr;
        ^
    
    AssertionError [ERR_ASSERTION]: ifError got unwanted exception: Command failed: ./node --help
    '.' is not recognized as an internal or external command,
    operable program or batch file.
    
        at ChildProcess.exithandler (child_process.js:282:12)
        at ChildProcess.emit (events.js:182:13)
        at maybeClose (internal/child_process.js:944:16)
        at Process.ChildProcess._handle.onexit (internal/child_process.js:233:5)
  ...
```

Trott commented:
> Test failing on Windows

@willhayslett Using `process.execPath` (as @richardlau suggests in a different comment) would seem to fix that.

willhayslett commented:
Thanks @richardlau and @Trott, changes committed.

vsemozhetbyt commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14256/

vsemozhetbyt commented:
Is one fail of `parallel/test-http-server-keep-alive-timeout` in one FreeBSD machine unrelated? Are we good to land?

Trott commented:
@vsemozhetbyt Sure seems like it ought to be unrelated, but no landing without a green CI everywhere, so here's a FreeBSD re-run: https://ci.nodejs.org/job/node-test-commit-freebsd/16998/

Trott commented:
This time a different test failed on FreeBSD, also seemingly unrelated.

Trying again: https://ci.nodejs.org/job/node-test-commit-freebsd/16999/

vsemozhetbyt commented:
Rerun CI is green. I will land soon if nobody objects.

Trott commented:
Last one was green. :+1:

vsemozhetbyt commented:
Squashed and landed in a43e8896f2c252ef30b4da08402badbe3a17124f

@willhayslett Thank you for all the heavy lifting!
Thanks to all reviewers for the time and toil.

willhayslett commented:
Woo hoo! Thanks, everyone!
